### PR TITLE
ARROW-1123: Make jemalloc the default allocator

### PR DESCRIPTION
**Pull request by xhochy**
Change-Id: Ib03392431851773df3b59b2c9a4d9a7bd672d2cb

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -938,8 +943,60 @@ set(ARROW_STATIC_PRIVATE_LINK_LIBS
...
+  if (NOT APPLE)
+      set(ARROW_JEMALLOC_STATIC_LINK_LIBS ${ARROW_JEMALLOC_STATIC_LINK_LIBS} ${PTHREAD_LIBRARY})
+  endif()
+
+  if (ARROW_JEMALLOC_USE_SHARED)
+    set(ARROW_JEMALLOC_SHARED_LINK_LIBS
+      jemalloc_shared
+      )
+  else()
+    if (CMAKE_COMPILER_IS_GNUCXX)
```
Comment (wesm) [inline cpp/CMakeLists.txt]: Does this impact clang on Linux?

**Review by wesm — APPROVED** [review]
+1 modulo comment re: clang on Linux

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (xhochy) [inline cpp/CMakeLists.txt]: I'm not sure about that. I think it mostly impacts glibc versions that are installed on CentOS 5 and Ubuntu 12.04. On these ancient systems, I don't think we could build arrow using clang, thus I would not further investigate if this is specific to glibc or glibc+gcc.

**Comment by wesm** [comment]
Looks like this may be failing for only cpplint

**Comment by xhochy** [comment]
Everything expect some internals in gtest are broken with jemalloc. At the moment I'm not sure if the culprit is `jemalloc` or `gtest` or `libc++`.

**Comment by xhochy** [comment]
Green build: https://travis-ci.org/xhochy/arrow/builds/245418595

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**